fabtests/prov/efa/src/multi_ep_stress.c

```diff
@@ -0,0 +1,1589 @@
+			usleep(1000);
+		}
+	}
+
+	printf("Sender %d: All cycles completed, total ops=%lu\n",
+	       ctx->worker_id, total_ops);
+
+out:
+	ctx->status.active = false;
+
+	// Wait for notification thread to finish
+	if (ctx->notification_thread) {
+		pthread_join(ctx->notification_thread, NULL);
+	}
+
+	return (void *) (intptr_t) ret;
+}
+
+static int notify_endpoint_update(struct receiver_context *ctx)
+{
+	struct ep_message msg;
+	msg.msg_type = MSG_TYPE_EP_UPDATE;
+	msg.info.worker_id = ctx->worker_id;
+
+	// Get endpoint address
+	msg.info.addr_len = MAX_EP_ADDR_LEN;
+	int ret = fi_getname(&ctx->common.ep->fid, msg.info.ep_addr,
+			     &msg.info.addr_len);
+	if (ret)
+		return ret;
+
+	// Fill RMA info
+	msg.info.rma.remote_addr = (uint64_t) ctx->rx_buf;
+	msg.info.rma.rkey = fi_mr_key(ctx->mr);
+	msg.info.rma.length = opts.transfer_size * topts.msgs_per_endpoint * ctx->num_senders;
+
+	// Send to all connected senders
+	for (int i = 0; i < ctx->num_senders; i++) {
+		if (ctx->control_socks[i] < 0)
+			continue;
+
+		size_t sent = 0;
+		while (sent < sizeof(msg)) {
+			ret = send(ctx->control_socks[i], (char *) &msg + sent,
+				   sizeof(msg) - sent, 0);
+
+			if (ret < 0) {
+				if (errno == EAGAIN || errno == EWOULDBLOCK) {
+					// Would block, try again after small delay
+					usleep(1000); // 1ms delay
+					continue;
+				} else if (errno == EINTR) {
+					// Interrupted by signal, retry
+					// immediately
+					continue;
+				} else {
+					// Real error
+					fprintf(stderr,
+						"Receiver %d: Failed to notify "
+						"sender %d: %s\n",
+						ctx->worker_id, i,
+						strerror(errno));
+					return -errno;
+				}
+			}
+			sent += ret;
+		}
+
+		if (topts.verbose)
+			printf("Receiver %d: Notified sender new EP\n",
+			       ctx->worker_id);
+	}
+
+	return 0;
+}
+
+static void *run_receiver_worker(void *arg)
+{
+	struct receiver_context *ctx = (struct receiver_context *) arg;
+	int ret = 0;
+	int cycle = 0;
+	int msg_per_ep_lifecyle =
+		topts.msgs_per_endpoint / topts.receiver_ep_recycling;
+	ctx->status.active = true;
+
+	if (topts.shared_cq)
+		ctx->common.cq = shared_rxcq;
+
+	if (topts.shared_av) {
+		ctx->common.av = shared_rxav;
+	} else {
+		av_attr.count = ctx->num_senders;
+	}
+
+	for (cycle = 0; cycle < topts.receiver_ep_recycling; cycle++) {
+		ret = setup_endpoint(&ctx->common);
+		if (ret)
+			break;
+
+		ctx->status.ep_generation++;
+		ctx->status.ep_status = EP_RECEIVING;
+		ctx->status.poll_cq = true; // rand() % 2;
+
+		// Notify sender of new endpoint
+		ret = notify_endpoint_update(ctx);
+		if (ret < 0) {
+			fprintf(stderr,
+				"Receiver %d: Failed to notify endpoint update "
+				"for cycle %d: %d\n",
+				ctx->worker_id, cycle + 1, ret);
+			cleanup_endpoint(&ctx->common);
+			continue;
+		}
+
+		// sleep random time up to 100ms to emulate the real workload
+		int sleep_time = rand() % 100000;
+		usleep(sleep_time);
+		printf("Receiver %d: Sleeping for %d microseconds\n",
```

Comment:
Would be helpful to print random seed somewhere to make test failures reproducable.

Reply:
Good idea! I added it and I am able to reproduce some failures